### PR TITLE
[cuDNN] cuDNN SDPA (Flash Attention) Backward

### PR DESCRIPTION
**Pull request by eqy**
#113713
currently passing trivial smoke tests but I just totally pattern-matched bits and pieces of the autograd defs

Will also collect benchmark data,

CC @drisspg 

cc @csarofeen @ptrblck @xwang233

### CODE REVIEW

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 1]
```diff
@@ -412,6 +542,96 @@ void run_cudnn_SDP_fprop(
...
+       dV,
+       dropoutseed,
+       dropoutoffset,
+       handle,
+       key.pod);
+   }
+   auto [mha_graph, Q, K, V, attn_scale, Seed, Offset, O, Do, Stats, Dq, Dk, Dv] = graph_and_tensors_backward_values;
+   std::unordered_map<std::shared_ptr<fe::graph::Tensor_attributes>, void*> variant_pack = {
+    // inputs
+   {Q, q.data_ptr()},
```
Comment (Skylion007) [inline aten/src/ATen/native/cudnn/MHA.cpp:609]: shouldn't some of these be `const_data_ptr`?

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 2]
```diff
@@ -146,6 +148,52 @@ std::tuple<Tensor, Tensor, Tensor> _flash_attention_backward(
...
+                        value /*const Tensor& v*/,
+                        out /*const Tensor& o*/,
+                        grad_out/*const Tensor& dO*/,
+                        logsumexp.unsqueeze(-1)/*const Tensor& softmaxstats*/,
+                        dq/*Tensor& dQ*/,
+                        dk/*Tensor& dK*/,
+                        dv/*Tensor& dV*/,
+                        philox_seed/*Tensor& dropoutseed*/,
+                        philox_offset/*Tensor& dropoutoffset*/);
+    return std::make_tuple(dq, dk, dv);
```
Comment (Skylion007) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:195]:
```suggestion
    return std::make_tuple(std::move(dq), std::move(dk), std::move(dv));
```

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 3]
```diff
@@ -327,6 +345,118 @@ auto build_graph_and_tensors(
...
+     DK->set_output(true).set_dim(std::vector<int64_t>(dK.sizes().begin(), dK.sizes().end())).set_stride(std::vector<int64_t>(dK.strides().begin(), dK.strides().end()));
+     DV->set_output(true).set_dim(std::vector<int64_t>(dV.sizes().begin(), dV.sizes().end())).set_stride(std::vector<int64_t>(dV.strides().begin(), dV.strides().end()));
+     AT_CUDNN_FRONTEND_CHECK(mha_graph->validate());
+     AT_CUDNN_FRONTEND_CHECK(mha_graph->build_operation_graph(handle));
+     AT_CUDNN_FRONTEND_CHECK(
+       mha_graph->create_execution_plans({fe::HeurMode_t::A}));
+     AT_CUDNN_FRONTEND_CHECK(mha_graph->check_support(handle));
+     AT_CUDNN_FRONTEND_CHECK(mha_graph->build_plans(handle));
+     return std::make_tuple(
+       mha_graph, Q, K, V, attn_scale, Seed, Offset, O, DO, STATS, DQ, DK, DV);
```
Comment (Skylion007) [inline aten/src/ATen/native/cudnn/MHA.cpp:456]: Alot of these should be std::move()

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -14620,11 +14620,16 @@
     CUDA: _scaled_dot_product_efficient_attention_backward_cuda
   tags: nondeterministic_seeded
 
-- func: _scaled_dot_product_cudnn_attention(Tensor query, Tensor key, Tensor value, float dropout_p=0.0, bool is_causal=False, bool return_debug_mask=False, *, float? scale=None) -> (Tensor output, Tensor logsumexp, Tensor philox_seed, Tensor philox_offset)
+- func: _scaled_dot_product_cudnn_attention(Tensor query, Tensor key, Tensor value, float dropout_p=0.0, bool is_causal=False, bool return_debug_mask=False, *, float? scale=None) -> (Tensor output, Tensor logsumexp, Tensor cum_seq_q, Tensor cum_seq_k, SymInt max_q, SymInt max_k, Tensor philox_seed, Tensor philox_offset, Tensor debug_attn_mask)
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:14623]: Why is this a special function? @jbschlosser and @andrewor14 have spent litteral months removing these cudnn variants for conv and batchnorm. I really don't think we should be doing the same again with sdpa...

Comment (eqy) [reply aten/src/ATen/native/native_functions.yaml:14694]: I think in this case it's because cuDNN is just a backend in the same way that e.g., flash, mem-efficient, and math are separate backends because the rules for dispatching between them aren't expected to overlap (e.g., cuDNN SDPA isn't expected to exactly cover the support matrix of flash or mem-efficient attention)

Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:14682]: Me and @albanD had a long discussion about this on Friday. Historically there has been a similar pattern for aten.op bloat followed by the need to unify a large number of backend behind composite explicit ops. The two exemplars of this are torch's cuDNN op, and batch norm with which consolidation is still ongoing by @andrewor14. 

The conclusion was that for now we are okay with paying the backend tech debt (added ops and added coverage surface) for the sake of velocity.

File: aten/src/ATen/native/cudnn/MHA.cpp (hunk 1 above)
Comment (eqy) [reply aten/src/ATen/native/cudnn/MHA.cpp]: Checked this again and it currently seems to be a limitation of cuDNN's `variantpack`, which only accepts `void *` pointers.

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 5]
```diff
@@ -324,7 +365,143 @@ auto build_graph_and_tensors(
...
+    const Tensor& dO,
+    const Tensor& softmaxstats,
+    Tensor& dQ,
+    Tensor& dK,
+    Tensor& dV,
+    const Tensor& dropoutseed,
+    const Tensor& dropoutoffset,
+    cudnnHandle_t& handle,
+    MHAParams& params) {
+  auto dtype = fe::DataType_t::HALF;
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:401]: does this support fp32?

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 6]
```diff
@@ -324,7 +365,143 @@ auto build_graph_and_tensors(
...
+  if (q.scalar_type() == kBFloat16) {
+    dtype = fe::DataType_t::BFLOAT16;
+  }
+  auto mha_graph = std::make_shared<fe::graph::Graph>();
+  mha_graph->set_io_data_type(dtype)
+      .set_intermediate_data_type(fe::DataType_t::FLOAT)
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:407]: Nit: maybe leave a comment at the top with some of the choices made about the graph construction

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 7]
```diff
@@ -324,7 +365,143 @@ auto build_graph_and_tensors(
...
+              std::vector<int64_t>(k.strides().begin(), k.strides().end())));
+  auto V = mha_graph->tensor(
+      fe::graph::Tensor_attributes()
+          .set_name("V")
+          .set_dim(std::vector<int64_t>(v.sizes().begin(), v.sizes().end()))
+          .set_stride(
+              std::vector<int64_t>(v.strides().begin(), v.strides().end())));
+  auto attn_scale =
+      mha_graph->tensor(fe::graph::Tensor_attributes()
+                            .set_name("attn_scale")
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:429]: does this support arbitrary attention bias tensors?

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 8]
```diff
@@ -324,7 +365,143 @@ auto build_graph_and_tensors(
...
+                                      .set_data_type(fe::DataType_t::INT32));
+  auto O = mha_graph->tensor(
+      fe::graph::Tensor_attributes()
+          .set_name("O")
+          .set_dim(std::vector<int64_t>(o.sizes().begin(), o.sizes().end()))
+          .set_stride(
+              std::vector<int64_t>(o.strides().begin(), o.strides().end())));
+  auto STATS = mha_graph->tensor(
+      fe::graph::Tensor_attributes()
+          .set_name("stats")
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:452]: nit: The softmaxstats is the logsumexp of the attention scores right? maybe softmaxstats is the nomenclature for cuDNN thats fine but if you wanted to align with the other kernels

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 9]
```diff
@@ -324,7 +365,143 @@ auto build_graph_and_tensors(
...
+              softmaxstats.strides().begin(), softmaxstats.strides().end()))
+          .set_data_type(fe::DataType_t::FLOAT));
+  auto DO = mha_graph->tensor(
+      fe::graph::Tensor_attributes()
+          .set_name("dO")
+          .set_dim(std::vector<int64_t>(dO.sizes().begin(), dO.sizes().end()))
+          .set_stride(
+              std::vector<int64_t>(dO.strides().begin(), dO.strides().end())));
+  auto sdpa_backward_options = fe::graph::SDPA_backward_attributes()
+                                   .set_name("flash_attention_backward")
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:465]: maybe different name? to remove some confusion between the existing flash impl

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 10]
```diff
@@ -412,6 +589,87 @@ void run_cudnn_SDP_fprop(
...
+                      // pass by value
+                      {attn_scale, &scaling_factor}};
+  if (dropout_probability != 0.0f) {
+    variant_pack[Seed] = dropoutseed.data_ptr();
+    variant_pack[Offset] = dropoutoffset.data_ptr();
+  }
+  auto workspace_size = mha_graph->get_workspace_size();
+  auto workspace_ptr =
+      c10::cuda::CUDACachingAllocator::get()->allocate(workspace_size);
+  TORCH_INTERNAL_ASSERT(!workspace_size || workspace_ptr.get());
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:667]: FYI that has bitten some people TORCH_INTERNAL_ASSERT is only active when building with debug.  This first assert probably makes sense to be internal, but do you think the execution of the graph should be a normal TORCH_CHECK ?

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu (hunk 2 above)
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:195]: std::move?

**Review by drisspg — COMMENTED** [review]
Looks good, mostly just small nits, and if you have any perf numbers that would be great!

**Comment by eqy** [comment]
Let me see if I can just rerun my existing forward script on forward + backward now...

**Comment by drisspg** [comment]
@eqy  The failure can be skipped by adding the op here: https://github.com/pytorch/pytorch/blob/c2f8bfae9cf8d3e3729ad8137d72a2c088c6a39c/test/forward_backward_compatibility/check_forward_backward_compatibility.py#L42

Since we dont actually dispatch to this op yet there shouldnt be any actual fc/bc concerns yet

**Comment by eqy** [comment]
Some quick perf numbers on H100:
```
-- | -- | --
0.0002416009665 | 0.0001961349626 | 1.231809787
0.001160450047 | 0.001005101018 | 1.154560613
0.0002051059855 | 0.0001553309849 | 1.320444763
0.001090625999 | 0.0009105839999 | 1.197721462
0.0003347190213 | 0.000274504011 | 1.219359309
0.001452401979 | 0.001198713959 | 1.211633491
0.0002948690089 | 0.0002105549793 | 1.400437121
0.001369567006 | 0.001101611007 | 1.243240124
0.0005467190058 | 0.0004442589707 | 1.230631325
0.002214480017 | 0.001602184027 | 1.382163334
0.0004940720391 | 0.0003251829767 | 1.519366248
0.002092602954 | 0.001431051001 | 1.46228398
0.001009774976 | 0.000770202023 | 1.311052096
0.003472366021 | 0.002579549968 | 1.346113107
0.0009175770101 | 0.0005585419713 | 1.642807627
0.003194732009 | 0.002253493003 | 1.417680022
0.001932165003 | 0.001437512052 | 1.344103516
0.005819034006 | 0.004001188965 | 1.454326216
0.001732694975 | 0.001016785041 | 1.704091724
0.005280279031 | 0.003570755012 | 1.478757017
0.003814590978 | 0.002746983024 | 1.388647453
0.01058057602 | 0.007186224975 | 1.472341328
0.003441496985 | 0.001926082012 | 1.786786318
0.009478844004 | 0.005942063988 | 1.595210692
0.01432821399 | 0.01049380499 | 1.365397394
0.03917209699 | 0.02671906003 | 1.466073168
0.01395738905 | 0.007517435995 | 1.856668824
0.03742853296 | 0.02278259303 | 1.642856584
0.05707501702 | 0.04218120797 | 1.353091098
0.15418196 | 0.107195058 | 1.43833086
0.05113227898 | 0.02938999602 | 1.739785162
0.146293823 | 0.08862046397 | 1.650790534
0.236646279 | 0.172920093 | 1.368529677
0.609717789 | 0.422645694 | 1.442621557
0.219155912 | 0.142931923 | 1.533288767
0.551255892 | 0.354812464 | 1.553654248
```

**Review by drisspg — APPROVED** [review]
Looks good! Thank you, the follow up PR for numerics testing should be easier to disentangle now!

**Comment by jeanschmidt** [comment]
@seemethere mentioned that we should protect like this:

```
+#ifdef __HIP_PLATFORM_AMD__
+#include <ATen/native/cudnn/hip/MHA.h>
+#else
 #include <ATen/native/cudnn/MHA.h>
+#endif
```

I don;t have lots of context here...

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 11]
```diff
@@ -43,6 +43,8 @@
 #include <ATen/native/transformers/cuda/mem_eff_attention/pytorch_utils.h>
 #endif
 
+#include <ATen/native/cudnn/MHA.h>
```
Comment (seemethere) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:46]:
```suggestion
#ifdef __HIP_PLATFORM_AMD__
#include <ATen/native/cudnn/hip/MHA.h>
#else
#include <ATen/native/cudnn/MHA.h>
#endif
```
